### PR TITLE
feat(auth): Add bulk get/delete methods

### PR DESCRIPTION
**Pull request by rsgowman**
This PR allows callers to retrieve a list of users by unique identifier (uid, email, phone, federated provider uid) as well as to delete a list of users.

RELEASE NOTE: Added `GetUsersAsync()` and `DeleteUsersAsync()` APIs for retrieving and deleting user accounts in bulk.

### CODE REVIEW

File: FirebaseAdmin/FirebaseAdmin/Auth/BatchDeleteResponse.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    /// </summary>
+    internal sealed class BatchDeleteResponse
+    {
+        /// <summary>
+        /// Gets or sets a list of errors.
+        /// </summary>
+        [JsonProperty(PropertyName = "errors")]
+        public List<ErrorInfo> Errors { get; set; }
+
+        internal sealed class ErrorInfo
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Auth/BatchDeleteResponse.cs]: Would it make sense to simply use the public `ErrorInfo` class here?

File: FirebaseAdmin/FirebaseAdmin/Auth/ProviderIdentifier.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        internal override void Populate(GetAccountInfoRequest payload)
+        {
+            payload.AddFederatedUserId(this.providerId, this.providerUid);
+        }
+
+        internal override bool Matches(UserRecord userRecord)
+        {
+            foreach (IUserInfo userInfo in userRecord.ProviderData)
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Auth/ProviderIdentifier.cs]: Use Linq

```
return userRecord.ProviderData.Any(...);
```

File: FirebaseAdmin/FirebaseAdmin/Auth/GetAccountInfoRequest.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+using System.Linq;
+
+namespace FirebaseAdmin.Auth
+{
+    /// <summary>
+    /// Represents a request to lookup account information.
+    /// </summary>
+    internal sealed class GetAccountInfoRequest
+    {
+        private IList<string> uids = null;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Auth/GetAccountInfoRequest.cs]: If you want you can preinitialize these to empty lists:

```
private readonly IList<string> uids = new List<string>();
```

Then check `!uids.Any()` in your Build methods.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseUserManager.cs [hunk 4]
```diff
@@ -158,6 +164,33 @@ internal static FirebaseUserManager Create(FirebaseApp app)
...
+        {
+            if (identifiers.Count == 0)
+            {
+                return new GetAccountInfoResponse();
+            }
+
+            if (identifiers.Count > MaxGetAccountsBatchSize)
+            {
+                throw new ArgumentException(
+                        "`identifier` parameter must have <= " + MaxGetAccountsBatchSize
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseUserManager.cs]: Use string interpolation whenever possible.

```$"`identifier` parameter must have <= {MaxGetAccountsBatchSize}..."```

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 5]
```diff
@@ -413,6 +414,77 @@ public async Task<UserRecord> GetUserByPhoneNumberAsync(string phoneNumber)
...
+        /// than 100 identifiers are specified.</exception>
+        public async Task<GetUsersResult> GetUsersAsync(
+            IReadOnlyCollection<UserIdentifier> identifiers, CancellationToken cancellationToken)
+        {
+            var userManager = this.IfNotDeleted(() => this.userManager.Value);
+
+            GetAccountInfoResponse response = await userManager.GetAccountInfoByIdentifiersAsync(identifiers, cancellationToken)
+                .ConfigureAwait(false);
+
+            IEnumerable<UserRecord> userRecords;
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Move all this logic into the constructor of `GetUsersResult`?

File: FirebaseAdmin/FirebaseAdmin.Tests/Auth/FirebaseUserManagerTest.cs [hunk 6]
```diff
@@ -326,6 +326,133 @@ public async Task GetUserByPhoneNumberEmpty()
...
+
+            var notFoundIds = new List<UserIdentifier>()
+            {
+                new UidIdentifier("id that doesnt exist"),
+            };
+
+            var getUsersResult = await auth.GetUsersAsync(notFoundIds);
+            Assert.Empty(getUsersResult.Users);
+            Assert.Single(getUsersResult.NotFound);
+            Assert.Equal(notFoundIds[0], getUsersResult.NotFound.First());
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin.Tests/Auth/FirebaseUserManagerTest.cs]:
```
Assert.Same(notFoundIds[0], getUsersResult.NotFound.Single()); 
```

File: FirebaseAdmin/FirebaseAdmin/Auth/UserRecordArgs.cs [hunk 7]
```diff
@@ -130,39 +120,87 @@ private static string CheckUid(string uid, bool required = false)
...
                 }
             }
+            else if (providerUid == string.Empty)
+            {
+                throw new ArgumentException(nameof(providerUid) + " must not be empty");
+            }
+        }
 
-            return phoneNumber;
+        internal CreateUserRequest ToCreateUserRequest()
```
Comment (hiranya911) [inline FirebaseAdmin/FirebaseAdmin/Auth/UserRecordArgs.cs]: Are these used somewhere?

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 5 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done. Also simplified (via linq) similar to your other suggestion.

File: FirebaseAdmin/FirebaseAdmin/Auth/BatchDeleteResponse.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace FirebaseAdmin.Auth
+{
+    /// <summary>
+    /// JSON data binding for BatchDeleteResponse messages sent by Google Identity Toolkit service.
```
Comment (egilmorez) [inline FirebaseAdmin/FirebaseAdmin/Auth/BatchDeleteResponse.cs:21]: Suggest backticks for BatchDeleteResponse.

File: FirebaseAdmin/FirebaseAdmin/Auth/DeleteUsersResult.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        /// <summary>
+        /// Gets the number of users that were deleted successfully (possibly zero). Users that
+        /// did not exist prior to calling
+        /// <see cref="M:FirebaseAuth.DeleteUsersAsync(IReadOnlyList{String})"/> are considered to
+        /// be successfully deleted.
+        /// </summary>
+        public int SuccessCount { get; }
+
+        /// <summary>
+        /// Gets the number of users that failed to be deleted (possibly zero).
```
Comment (egilmorez) [inline FirebaseAdmin/FirebaseAdmin/Auth/DeleteUsersResult.cs:62]: I feel like we discussed this wording in a CL, but I can't remember the wording we settled on :(

Best way is probably to repeat the method name:

"Gets the number of users that `<method_name>` failed to delete (possibly zero).

File: FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace FirebaseAdmin.Auth
+{
+    // TODO(rsgowman): This class is expected to also be used for the
+    // ImportUsersAsync() method... once that exists.
+
+    /// <summary>
+    /// Represents an error encountered while delete users via the
```
Comment (egilmorez) [inline FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs:21]: I think this should be "while deleting"

File: FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    public sealed class ErrorInfo
+    {
+        internal ErrorInfo(int index, string reason)
+        {
+            this.Index = index;
+            this.Reason = reason;
+        }
+
+        /// <summary>
+        /// Gets the index of the failed user in the list passed to the
```
Comment (egilmorez) [inline FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs:33]: Is there a better term than "failed user" that we can use?

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 12]
```diff
@@ -603,6 +656,70 @@ await userManager.DeleteUserAsync(uid, cancellationToken)
                 .ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Deletes the users specified by the given identifiers.
+        /// <para>
+        /// Deleting a non-existing user won't generate an error. (i.e. this method is idempotent.)
+        /// Non-existing users will be considered to be successfully deleted, and will therefore be
+        /// counted in the DeleteUserResult.SuccessCount value.
```
Comment (egilmorez) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:664]: Suggest backticks for DeleteUserResult.SuccessCount, here and similar copies.

File: FirebaseAdmin/FirebaseAdmin/Auth/GetAccountInfoRequest.cs [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace FirebaseAdmin.Auth
+{
+    /// <summary>
+    /// Represents a request to lookup account information.
```
Comment (egilmorez) [inline FirebaseAdmin/FirebaseAdmin/Auth/GetAccountInfoRequest.cs:21]: Suggest "look up" unless that is truly inaccurate.

File: FirebaseAdmin/FirebaseAdmin/Auth/BatchDeleteResponse.cs (hunk 8 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/BatchDeleteResponse.cs]: Done.

File: FirebaseAdmin/FirebaseAdmin/Auth/DeleteUsersResult.cs (hunk 9 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/DeleteUsersResult.cs]: Done.

File: FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs (hunk 10 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs]: Done.

File: FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs (hunk 11 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/ErrorInfo.cs]: It's a bit wordy, but how about:

"Gets the index of the user that was unable to be deleted in the list passed to the DeleteUsersAsync method."
?

(I've made this change. But feel free to improve it further!)

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 12 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done.

File: FirebaseAdmin/FirebaseAdmin/Auth/GetAccountInfoRequest.cs (hunk 13 above)
Comment (rsgowman) [reply FirebaseAdmin/FirebaseAdmin/Auth/GetAccountInfoRequest.cs]: Done.
